### PR TITLE
Composition navigation properties and composition child entity sets

### PR DESCRIPTION
**Pull request by ralfhandl**
Discuss:
- [x] is the annotation `CompositionChild` redundant as it can be determined from the navigation property bindings (association sets in OData V2) and annotation `Composition` on the navigation property?
- [x] is the partner navigation property required? No.

### CODE REVIEW

**Review by RaizoC — APPROVED** [review]

**Review by RaizoC — APPROVED** [review]

**Review by RaizoC — APPROVED** [review]

File: vocabularies/Common.xml [hunk 1]
```diff
@@ -569,6 +569,22 @@ The minimum structure is
...
+        </Annotation>
+      </Term>
+
+      <Term Name="CompositionChild" Type="Core.Tag" Nullable="false" DefaultValue="true" AppliesTo="EntitySet">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="The entities in this set are composition children and have an existential dependency on their composition parent" />
+        <Annotation Term="Core.LongDescription">
+          <String>The navigation property representing logical composition from the "parent" entity type to the "child" entity type is annotated with [Common.Composition](#Composition)</String>
+        </Annotation>
+      </Term>
```
Comment (HeikoTheissen) [inline vocabularies/Common.xml:586]:
```suggestion
        <Annotation Term="Core.LongDescription">
          <String>The entities related via this navigation property have an existential dependency on their composition parent. The entity set of the composition parent MUST contain a NavigationPropertyBinding for this navigation property.</String>
        </Annotation>
      </Term>

```

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

File: vocabularies/Common.xml [hunk 2]
```diff
@@ -569,6 +569,22 @@ The minimum structure is
         </Property>
       </ComplexType>
 
+      <Term Name="Composition" Type="Core.Tag" Nullable="false" DefaultValue="true" AppliesTo="NavigationProperty">
+        <Annotation Term="Common.Experimental" />
+        <Annotation Term="Core.Description" String="The annotated non-containment navigation property represents a logical composition" />
```
Comment (HeikoTheissen) [inline vocabularies/Common.xml:574]:
```suggestion
        <Annotation Term="Core.Description" String="The annotated navigation property represents a logical composition, even though it is non-containment" />
```

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

**Review by HeikoTheissen — DISMISSED** [review]

File: vocabularies/Common.xml [hunk 3]
```diff
@@ -569,6 +569,15 @@ The minimum structure is
         </Property>
       </ComplexType>
 
+      <Term Name="Composition" Type="Core.Tag" Nullable="false" DefaultValue="true" AppliesTo="NavigationProperty">
```
Comment (HeikoTheissen) [inline vocabularies/Common.xml:572]:
```suggestion
      <Term Name="Composition" Type="Core.Tag" Nullable="false" DefaultValue="true" AppliesTo="NavigationProperty">
        <Annotation Term="Core.AppliesViaContainer"/>
```
Do we want to add this?

Comment (ralfhandl) [reply vocabularies/Common.xml:572]: No, I wanted to directly annotate the navigation properties, see the example.

File: examples/Common.Composition-sample.xml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        </Key>
+        <Property Name="SalesOrder" Type="Edm.String" Nullable="false" MaxLength="10" />
+        <Property Name="SalesOrderItem" Type="Edm.String" Nullable="false" MaxLength="6" />
+        <!-- ... and a lot more -->
+        <NavigationProperty Name="_SalesOrder" Type="this.SalesOrder_Type" Nullable="false" Partner="_Item">
+          <ReferentialConstraint Property="SalesOrder" ReferencedProperty="SalesOrder" />
+        </NavigationProperty>
+      </EntityType>
+
+      <Annotations Target="this.SalesOrder/_Item">
```
Comment (HeikoTheissen) [inline examples/Common.Composition-sample.xml:39]: Invalid target, what do you mean?
* The entity type `this.SalesOrder_Type/_Item`
* Or the entity set `this.container.SalesOrder/_Item`

**Review by HeikoTheissen — CHANGES_REQUESTED** [review]

File: examples/Common.Composition-sample.xml (hunk 4 above)
Comment (ralfhandl) [reply examples/Common.Composition-sample.xml]: Type, fixed

**Review by HeikoTheissen — APPROVED** [review]

**Comment by HeikoTheissen** [comment]
Do we require that a `Composition` navprop has a partner navprop? Must a partner navprop also be accompanied by a navprop binding?

**Comment by ralfhandl** [comment]
_Ideally_ it has a partner, and ideally the partner has referential constraints to identify the parts of the key that are already determined by the composition parent.

I'll have to write a small PoC to figure out whether we can work without this additional information.

**Pull request merged**